src/precice/bindings/matlab/+precice/@SolverInterface/private/preciceGateway.cpp

```diff
@@ -0,0 +1,397 @@
+// Gateway MexFunction object
+#include "mex.hpp"
+#include "mexAdapter.hpp"
+#include <iostream>
+#include <sstream>
+#include "precice/SolverInterface.hpp"
+
+using namespace matlab::data;
+using matlab::mex::ArgumentList;
+using namespace precice;
+using namespace precice::constants;
+
+class MexFunction: public matlab::mex::Function {
+private:
+    SolverInterface* interface;
+    ArrayFactory factory;
+    bool constructed;
+    std::shared_ptr<matlab::engine::MATLABEngine> matlabPtr = getEngine();
+    
+    void myMexPrint(const std::string text) {
+        matlabPtr->feval(u"fprintf",0,std::vector<Array>({factory.createScalar(text)}));
+    }
+    
+public:
+    MexFunction(): constructed{false}, factory{}, interface{NULL} {}
+
+    void operator()(ArgumentList outputs, ArgumentList inputs) {
+        bool valid=false;
+        // define the functionID
+        const TypedArray<uint8_t> functionIDArray = inputs[0];
+        int functionID = functionIDArray[0];
+        
+        // Debug output
+        // std::cout << "Gateway function " << (int)functionID << " was called." << std::endl;
+        
+        // Abort if constructor was not called before, or if constructor 
+        // was called on an existing solverInterface
+        if (functionID==0 && constructed) {
+            myMexPrint("Constructor was called but interface is alread construced.");
+            return;
+        }
+        if (!constructed && functionID!=0) {
+            myMexPrint("Interface was not construced before.");
+            return;
+        }
+        
+        switch (functionID) {
```

Comment:
Put the function ids in a `enum class FunctionID { Constructor, Destructor, ... } `.
Then you can `static_cast<FunctionID>(functionID)` and use that in the `switch(...)`.
This makes the following `case(...)` statements more explicit and less error prone: `case FunctionID::Constructor`.

Use `return` in the cases instead of `break` as there is no error handling.